src/Units/WeaponModel.java

```diff
@@ -0,0 +1,75 @@
+package Units;
+
+import java.io.Serializable;
+
+import Units.AWBWUnits.AWBWUnitModel;
+import Units.DoRUnits.DoRUnitModel;
+
+public abstract class WeaponModel implements Serializable
+{
+  private static final long serialVersionUID = 1L;
+
+  public boolean canFireAfterMoving;
+  public boolean hasInfiniteAmmo;
+  public int minRange;
+  public int maxRange;
+
+  protected WeaponModel(boolean infiniteAmmo, int minRange, int maxRange)
+  {
+    hasInfiniteAmmo = infiniteAmmo;
+    if( minRange > 1 )
+    {
+      canFireAfterMoving = false;
+    }
+    else
+    {
+      canFireAfterMoving = true;
+    }
+    this.minRange = minRange;
+    this.maxRange = maxRange;
+  }
+  protected WeaponModel(boolean infiniteAmmo)
+  {
+    this(infiniteAmmo, 1, 1);
+  }
+  protected WeaponModel()
+  {
+    this(true, 1, 1);
+  }
+  public abstract WeaponModel clone();
+
+  public boolean loaded(Unit user)
+  {
+    if( hasInfiniteAmmo )
+      return true;
+    else
+      return user.ammo > 0;
+  }
+
+  /**
+   * @return returns its base damage against defender if the unit is in range
+   */
+  public double getDamage(UnitModel defender, int range)
+  {
+    if( (range >= minRange) && (range <= maxRange) )
+      return getDamage(defender);
+    return 0;
+  }
+  /**
+   * @return returns its base damage against that unit type
+   */
+  public double getDamage(UnitModel defender)
+  {
+    if( defender == null )
+      throw new IllegalArgumentException("Stare not into the void. It just may stare back.");
```

Comment:
Now you're just trying to get a reaction. You know what? It's a unique message that we can use it to find the problem. I'm fine with this. I probably shouldn't do code review past midnight.